src/crystal/system/unix/io_uring.cr

```diff
@@ -0,0 +1,428 @@
+  # submit a chain of SQE at once. Submits pending SQE and waits if needed.
+  def reserve(count : Int32) : Nil
+    raise ArgumentError.new("Can't reserve more SQE than available in the SQ ring") if count > @sq_entries.value
+
+    loop do
+      tail = @sq_tail # Atomic::Ops.load(@sq_ktail, :monotonic, volatile: true)
+      head = Atomic::Ops.load(@sq_khead, :monotonic, volatile: true)
+      size = tail &- head
+
+      if (@sq_entries.value - size) >= count
+        break
+      else
+        submit(wait: true)
+      end
+    end
+  end
+
+  # Reserves a slot and returns the next SQE slot in the SQ ring. Submits
+  # pending submissions and waits if the ring is full.
+  def next_sqe : LibC::IoUringSqe*
+    reserve(1)
+    unsafe_next_sqe
+  end
+
+  # Identical to `#next_sqe` but also returns a stable pointer to a
+  # `LibC::Timespec` struct to associate to the SQE.
+  #
+  # This might be required with SQPOLL or when the ring doesn't support the
+  # IORING_FEAT_SUBMIT_STABLE feature.
+  def next_sqe_with_timespec : {LibC::IoUringSqe*, LibC::Timespec*}
+    reserve(1)
+    unsafe_next_sqe_with_timespec
+  end
+
+  # WARNING: must call `#reserve` before calling `#unsafe_next_sqe`!
+  def unsafe_next_sqe : LibC::IoUringSqe*
+    index = @sq_tail & @sq_mask.value
+    @sq_tail &+= 1
+
+    sqe = @sqes.as(LibC::IoUringSqe*) + index
+    LibIntrinsics.memset(sqe, 0_u8, sizeof(LibC::IoUringSqe), false)
+
+    sqe
+  end
+
+  # WARNING: must call `#reserve` before calling `#unsafe_next_sqe_with_timespec`!
+  def unsafe_next_sqe_with_timespec : {LibC::IoUringSqe*, LibC::Timespec*}
+    index = @sq_tail & @sq_mask.value
+    @sq_tail &+= 1
+
+    sqe = @sqes.as(LibC::IoUringSqe*) + index
+    LibIntrinsics.memset(sqe, 0_u8, sizeof(LibC::IoUringSqe), false)
+
+    {sqe, @timespecs + index}
+  end
+
+  # Submit pending SQE in the SQ ring if needed. Wake SQPOLL thread if sleeping.
+  # Blocks until at least one SQE becomes available when *wait* is true.
+  def submit(flags : UInt32 = 0_u32, wait : Bool = false)
+    # make new tail and previous writes visible to the kernel threads
+    Atomic::Ops.store(@sq_ktail, @sq_tail, :release, volatile: true)
+
+    if sq_poll?
+      if wait
+        flags |= LibC::IORING_ENTER_SQ_WAIT
+      elsif sq_need_wakeup?
+        flags |= LibC::IORING_ENTER_SQ_WAKEUP
+      else
+        return
+      end
+    end
+
+    loop do
+      head = Atomic::Ops.load(@sq_khead, :monotonic, volatile: true)
+      to_submit = @sq_tail &- head
+
+      ret = enter(to_submit, flags: flags)
+      break unless ret == -LibC::EINTR && (!sq_poll? || sq_need_wakeup?)
+    end
+  end
+
+  # Call `io_uring_enter` syscall. Panics on EBADR (can't recover from lost
+  # CQE), returns -EINTR or -EBUSY, and raises on other errnos, otherwise
+  # returns the int returned by the syscall.
+  def enter(to_submit : UInt32 = 0, min_complete : UInt32 = 0, flags : UInt32 = 0) : Int32
```

Comment:
By the way, one thing that has changed with regards to external libraries is that liburing has changed so that in addition to the old .so  that had a million functions missing due to them being declared as `static inline` and therefore requiring a file with c shims to be used, they now also provide a liburing-ffi.so, which has dropped those static inline. This due to rust people wanting to use the library, and they have mostly the same issues with `static inline` as we do even if they are a bit further along. So that is a lot less of a hassle than it used to be.

So if we wanted we could exchange a whole lot of the complexity in this file in exchange for linking against liburing-ffi. This seems complex enough that it may be worth the trouble of extra dependencies.

Reply:
Sadly that would require a specific version that won't be available for the years to come in distributions, and it's not installed by default (unlike libc)... unless we build and distribute our own copy (nope).

Each prep method also calls a bunch of other prep methods, so we'd lose the benefit of inlining all the SQE setup.

It's also not that complex (setup is textbook boilerplate for example) though I admit I shall check the memory ordering of atomics again :sweat_smile: